### ISSUE TITLE
support for htslib 1.16

### ISSUE THREAD

**Issue by emollier** [pysam-developers/pysam#1131]
Greetings,

we recently bumped bcftools, htslib and samtools to version 1.16 in Debian Sid, and we noticed the test suite of pysam is failing 10 items with the new packages versions.  A full test log is currently available on [Debian CI infrastructure](https://ci.debian.net/data/autopkgtest/testing/amd64/p/python-pysam/26567166/log.gz).  An excerpt with the relevant test errors is available below:
```
=================================== FAILURES ===================================
______________________ TestTruncatedBAM.testTruncatedBam2 ______________________

self = <AlignmentFile_test.TestTruncatedBAM testMethod=testTruncatedBam2>

    def testTruncatedBam2(self):
>       self.assertRaises(IOError,
                          pysam.AlignmentFile,
                          os.path.join(BAM_DATADIR, 'ex2_truncated.bam'))
E       AssertionError: OSError not raised by AlignmentFile

tests/AlignmentFile_test.py:1444: AssertionError
__________________ TestTruncatedBAM.testTruncatedBamIterator ___________________

self = <AlignmentFile_test.TestTruncatedBAM testMethod=testTruncatedBamIterator>

    def testTruncatedBamIterator(self):
        s = pysam.AlignmentFile(os.path.join(BAM_DATADIR, 'ex2_truncated.bam'),
                                ignore_truncation=True)
    
        def iterall(x):
            return len([a for a in x])
>       self.assertRaises(IOError, iterall, s)
E       AssertionError: OSError not raised by iterall

tests/AlignmentFile_test.py:1454: AssertionError
___________________ TestVCFFromVariantFile_20.testConnecting ___________________

>   self.header = makeVariantHeader(hdr)

pysam/libcbcf.pyx:4292: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError('cannot create VariantHeader')
E   ValueError: cannot create VariantHeader

pysam/libcbcf.pyx:2193: ValueError

During handling of the above exception, another exception occurred:

self = <tabix_test.TestVCFFromVariantFile_20 testMethod=testConnecting>

    def testConnecting(self):
    
        if self.get_failure_stage() == "opening":
            self.assertRaises(ValueError,
                              self.open_vcf,
                              self.tmpfilename + ".gz")
        else:
>           self.open_vcf(self.tmpfilename + ".gz")

tests/tabix_test.py:748: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/tabix_test.py:1010: in open_vcf
    with pysam.VariantFile(fn) as inf:
pysam/libcbcf.pyx:4054: in pysam.libcbcf.VariantFile.__init__
    self.open(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError('file `{}` does not have valid header (mode=`{}`) - is it VCF/BCF format?'.format(filename, mode))
E   ValueError: file `b'/tmp/autopkgtest-lxc.f40307db/downtmp/build.sWj/src/tests/tmp/pysamtests_tmp_setUp_le4khaxbvcf.gz'` does not have valid header (mode=`b'r'`) - is it VCF/BCF format?

pysam/libcbcf.pyx:4294: ValueError
____________________ TestVCFFromVariantFile_20.testParsing _____________________

>   self.header = makeVariantHeader(hdr)

pysam/libcbcf.pyx:4292: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError('cannot create VariantHeader')
E   ValueError: cannot create VariantHeader

pysam/libcbcf.pyx:2193: ValueError

During handling of the above exception, another exception occurred:

self = <tabix_test.TestVCFFromVariantFile_20 testMethod=testParsing>

    def testParsing(self):
    
        if self.get_failure_stage() in ("opening", "parsing"):
            return
    
>       itr = self.get_iterator()

tests/tabix_test.py:781: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/tabix_test.py:1003: in get_iterator
    self.vcf = pysam.VariantFile(self.filename)
pysam/libcbcf.pyx:4054: in pysam.libcbcf.VariantFile.__init__
    self.open(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError('file `{}` does not have valid header (mode=`{}`) - is it VCF/BCF format?'.format(filename, mode))
E   ValueError: file `b'/tmp/autopkgtest-lxc.f40307db/downtmp/build.sWj/src/tests/tabix_data/vcf/20.vcf'` does not have valid header (mode=`b'r'`) - is it VCF/BCF format?

pysam/libcbcf.pyx:4294: ValueError
___________________ TestVCFFromVariantFile_2.testConnecting ____________________

>   self.header = makeVariantHeader(hdr)

pysam/libcbcf.pyx:4292: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError('cannot create VariantHeader')
E   ValueError: cannot create VariantHeader

pysam/libcbcf.pyx:2193: ValueError

During handling of the above exception, another exception occurred:

self = <tabix_test.TestVCFFromVariantFile_2 testMethod=testConnecting>

    def testConnecting(self):
    
        if self.get_failure_stage() == "opening":
            self.assertRaises(ValueError,
                              self.open_vcf,
                              self.tmpfilename + ".gz")
        else:
>           self.open_vcf(self.tmpfilename + ".gz")

tests/tabix_test.py:748: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/tabix_test.py:1010: in open_vcf
    with pysam.VariantFile(fn) as inf:
pysam/libcbcf.pyx:4054: in pysam.libcbcf.VariantFile.__init__
    self.open(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError('file `{}` does not have valid header (mode=`{}`) - is it VCF/BCF format?'.format(filename, mode))
E   ValueError: file `b'/tmp/autopkgtest-lxc.f40307db/downtmp/build.sWj/src/tests/tmp/pysamtests_tmp_setUp_nuctzzofvcf.gz'` does not have valid header (mode=`b'r'`) - is it VCF/BCF format?

pysam/libcbcf.pyx:4294: ValueError
_____________________ TestVCFFromVariantFile_2.testParsing _____________________

>   self.header = makeVariantHeader(hdr)

pysam/libcbcf.pyx:4292: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError('cannot create VariantHeader')
E   ValueError: cannot create VariantHeader

pysam/libcbcf.pyx:2193: ValueError

During handling of the above exception, another exception occurred:

self = <tabix_test.TestVCFFromVariantFile_2 testMethod=testParsing>

    def testParsing(self):
    
        if self.get_failure_stage() in ("opening", "parsing"):
            return
    
>       itr = self.get_iterator()

tests/tabix_test.py:781: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/tabix_test.py:1003: in get_iterator
    self.vcf = pysam.VariantFile(self.filename)
pysam/libcbcf.pyx:4054: in pysam.libcbcf.VariantFile.__init__
    self.open(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError('file `{}` does not have valid header (mode=`{}`) - is it VCF/BCF format?'.format(filename, mode))
E   ValueError: file `b'/tmp/autopkgtest-lxc.f40307db/downtmp/build.sWj/src/tests/tabix_data/vcf/2.vcf'` does not have valid header (mode=`b'r'`) - is it VCF/BCF format?

pysam/libcbcf.pyx:4294: ValueError
___________________ TestVCFFromVariantFile_16.testConnecting ___________________

>   self.header = makeVariantHeader(hdr)

pysam/libcbcf.pyx:4292: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError('cannot create VariantHeader')
E   ValueError: cannot create VariantHeader

pysam/libcbcf.pyx:2193: ValueError

During handling of the above exception, another exception occurred:

self = <tabix_test.TestVCFFromVariantFile_16 testMethod=testConnecting>

    def testConnecting(self):
    
        if self.get_failure_stage() == "opening":
            self.assertRaises(ValueError,
                              self.open_vcf,
                              self.tmpfilename + ".gz")
        else:
>           self.open_vcf(self.tmpfilename + ".gz")

tests/tabix_test.py:748: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/tabix_test.py:1010: in open_vcf
    with pysam.VariantFile(fn) as inf:
pysam/libcbcf.pyx:4054: in pysam.libcbcf.VariantFile.__init__
    self.open(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError('file `{}` does not have valid header (mode=`{}`) - is it VCF/BCF format?'.format(filename, mode))
E   ValueError: file `b'/tmp/autopkgtest-lxc.f40307db/downtmp/build.sWj/src/tests/tmp/pysamtests_tmp_setUp_rebjrotuvcf.gz'` does not have valid header (mode=`b'r'`) - is it VCF/BCF format?

pysam/libcbcf.pyx:4294: ValueError
____________________ TestVCFFromVariantFile_16.testParsing _____________________

>   self.header = makeVariantHeader(hdr)

pysam/libcbcf.pyx:4292: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError('cannot create VariantHeader')
E   ValueError: cannot create VariantHeader

pysam/libcbcf.pyx:2193: ValueError

During handling of the above exception, another exception occurred:

self = <tabix_test.TestVCFFromVariantFile_16 testMethod=testParsing>

    def testParsing(self):
    
        if self.get_failure_stage() in ("opening", "parsing"):
            return
    
>       itr = self.get_iterator()

tests/tabix_test.py:781: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/tabix_test.py:1003: in get_iterator
    self.vcf = pysam.VariantFile(self.filename)
pysam/libcbcf.pyx:4054: in pysam.libcbcf.VariantFile.__init__
    self.open(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError('file `{}` does not have valid header (mode=`{}`) - is it VCF/BCF format?'.format(filename, mode))
E   ValueError: file `b'/tmp/autopkgtest-lxc.f40307db/downtmp/build.sWj/src/tests/tabix_data/vcf/16.vcf'` does not have valid header (mode=`b'r'`) - is it VCF/BCF format?

pysam/libcbcf.pyx:4294: ValueError
___________________ TestVCFFromVariantFile_23.testConnecting ___________________

>   self.header = makeVariantHeader(hdr)

pysam/libcbcf.pyx:4292: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError('cannot create VariantHeader')
E   ValueError: cannot create VariantHeader

pysam/libcbcf.pyx:2193: ValueError

During handling of the above exception, another exception occurred:

self = <tabix_test.TestVCFFromVariantFile_23 testMethod=testConnecting>

    def testConnecting(self):
    
        if self.get_failure_stage() == "opening":
            self.assertRaises(ValueError,
                              self.open_vcf,
                              self.tmpfilename + ".gz")
        else:
>           self.open_vcf(self.tmpfilename + ".gz")

tests/tabix_test.py:748: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/tabix_test.py:1010: in open_vcf
    with pysam.VariantFile(fn) as inf:
pysam/libcbcf.pyx:4054: in pysam.libcbcf.VariantFile.__init__
    self.open(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError('file `{}` does not have valid header (mode=`{}`) - is it VCF/BCF format?'.format(filename, mode))
E   ValueError: file `b'/tmp/autopkgtest-lxc.f40307db/downtmp/build.sWj/src/tests/tmp/pysamtests_tmp_setUp_yscuox5fvcf.gz'` does not have valid header (mode=`b'r'`) - is it VCF/BCF format?

pysam/libcbcf.pyx:4294: ValueError
____________________ TestVCFFromVariantFile_23.testParsing _____________________

>   self.header = makeVariantHeader(hdr)

pysam/libcbcf.pyx:4292: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError('cannot create VariantHeader')
E   ValueError: cannot create VariantHeader

pysam/libcbcf.pyx:2193: ValueError

During handling of the above exception, another exception occurred:

self = <tabix_test.TestVCFFromVariantFile_23 testMethod=testParsing>

    def testParsing(self):
    
        if self.get_failure_stage() in ("opening", "parsing"):
            return
    
>       itr = self.get_iterator()

tests/tabix_test.py:781: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/tabix_test.py:1003: in get_iterator
    self.vcf = pysam.VariantFile(self.filename)
pysam/libcbcf.pyx:4054: in pysam.libcbcf.VariantFile.__init__
    self.open(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError('file `{}` does not have valid header (mode=`{}`) - is it VCF/BCF format?'.format(filename, mode))
E   ValueError: file `b'/tmp/autopkgtest-lxc.f40307db/downtmp/build.sWj/src/tests/tabix_data/vcf/23.vcf'` does not have valid header (mode=`b'r'`) - is it VCF/BCF format?

pysam/libcbcf.pyx:4294: ValueError
=============================== warnings summary ===============================
../../../../../usr/lib/python3/dist-packages/_pytest/config/__init__.py:1252
  /usr/lib/python3/dist-packages/_pytest/config/__init__.py:1252: PytestConfigWarning: Unknown config option: pep8ignore
  
    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

../../../../../usr/lib/python3/dist-packages/_pytest/config/__init__.py:1252
  /usr/lib/python3/dist-packages/_pytest/config/__init__.py:1252: PytestConfigWarning: Unknown config option: pep8maxlinelength
  
    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

../../../../../usr/lib/python3/dist-packages/pyximport/pyximport.py:51
  /usr/lib/python3/dist-packages/pyximport/pyximport.py:51: DeprecationWarning: the imp module is deprecated in favour of importlib and slated for removal in Python 3.12; see the module's documentation for alternative uses
    import imp

tests/_compile_test.pyxbld:3
  /tmp/autopkgtest-lxc.f40307db/downtmp/build.sWj/src/tests/_compile_test.pyxbld:3: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
    from distutils.extension import Extension

../../../../../usr/lib/python3.10/distutils/command/build_ext.py:13
  /usr/lib/python3.10/distutils/command/build_ext.py:13: DeprecationWarning: The distutils.sysconfig module is deprecated, use sysconfig instead
    from distutils.sysconfig import customize_compiler, get_python_version

../../../../../usr/lib/python3/dist-packages/Cython/Compiler/Main.py:369
  /usr/lib/python3/dist-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: /tmp/autopkgtest-lxc.f40307db/downtmp/build.sWj/src/tests/_compile_test.pyx
    tree = Parsing.p_module(s, pxd, full_module_name)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
```

Have a nice day,  :)
Étienne.

**Comment by jmarshall** [comment]
You don't say what version of pysam you are trying to build. However according to your linked full log, it is 0.17.0.

The failures you report are test data issues that have already been fixed:

* The `TestVCFFromVariantFile` failures are due to stricter `#CHROM` parsing in HTSlib 1.14 and later. The test data was corrected in pysam 0.18.0 (see d46a82cc89bf687820c6404e11b4ea1094720341 and PR #1054).

* The `TestTruncatedBAM` failures are due to preparing the test data with a `samtools` built to use libdeflate. This part is a duplicate of #1128 and is fixed on **master** by 56b1624c21b02762fdcea3aa5f657a4152ca178e.

**Comment by emollier** [comment]
Hi John,

Thanks for the pointers, they helped me finish the pysam update to 0.19.1 in debian sid.  Sorry for my noise caused by my lousy search in the issue tracker.

Have a nice day,  :)
Étienne.
